src/extensionApi.ts

```diff
@@ -82,38 +84,36 @@ export class CommandHandler {
     }
 
     public async debugServer(context?: Protocol.ServerState): Promise<Protocol.StartServerResponse> {
-
         if (context === undefined) {
             const selectedServerId = await this.selectServer('Select server to start.');
             if (!selectedServerId) return;
-            context = this.serversData.serverStatus.get(selectedServerId);
+            context = this.explorer.serverStatus.get(selectedServerId);
         }
 
-        const debugInfo = await this.serversData.retrieveDebugInfo(context.server);
-
+        const debugInfo: DebugInfo = await DebugInfoProvider.retrieve(context.server, this.client);
         const extensionIsRequired = await this.checkExtension(debugInfo);
 
         if (extensionIsRequired) {
             vscode.window.showErrorMessage(extensionIsRequired);
             return;
         }
 
-        return this.startServer('debug', context).then(
-            status => {
-                vscode.debug.startDebugging(undefined,
-                    {
-                        type: 'java',
-                        request: 'attach',
-                        name: 'Debug (Remote)',
-                        hostName: 'localhost',
-                        port: `${status.details.properties['debug.details.port']}`
-                    }
-                );
-
-                return status;
-            }
-        );
-
+        this.startServer('debug', context)
+            .then(serverStarted => {
+                if (!serverStarted
+                    || !serverStarted.details) {
+                    return Promise.reject(`Failed to start server ${context.server.id}`);
+                }
+                const debugConfig: vscode.DebugConfiguration = {
+                    type: 'java',
+                    request: 'attach',
+                    name: 'Debug (Remote)',
+                    hostName: 'localhost',
+                    port: DebugInfoProvider.create(serverStarted.details).getPort()
+                };
```

Comment:
wouldn't it be better if we put the creation of the entire debugConfiguration inside DebugInfoProvider? I just thought that in future we could deal with different languages and it would be better if they are managed externally. What do you think?

Reply:
hmm, not sure about this. I would keep the debugInfo focused on parsing the server debug info (code cohesion). Maybe we should extract this to a new componet that provides utilities for dealing with the vscode debugger?

Reply:
In my idea there was a sort of factory which returned a debugConfiguration based on the necessity and the requests made by the user. Currently we just support java so maybe there is no need for such a thing. Btw yeah a new component it could be great for code organization

Reply:
@lstocchi good idea. Yes, I agree, let's make this factory once we need it (when we support other languages)